### DISCUSSION TITLE
Question: A step by step guide how to compile ISO

### DISCUSSION THREAD

**Discussion by CoodyCZ** [PartialVolume/shredos.x86_64#83] [Q&A]
@PartialVolume hi! Your work is great! And i'm a newbie in compiling... is there any way, to get from you a step by step guide, what must be instaled on pc, and how to compile ISO? Thanks!

**Comment by PartialVolume** [comment]
Thanks, much appreciated. Sorry this is a long post !

Yes, I'll see if I can put something together on how to build shredos, including an explanation of the key folders and files that differentiate shredos from a vanilla buildroot. How you include software that isn't available in buildroot, such as the master build of nwipe. Once you have a working setup it's literally 3 commands and a couple of hours work for your computer to build shredos finishing up with the .img or .iso. Building shredos on a fresh vanilla release of buildroot is something else altogether.

The way I'll do this is to create a document folder in shredos and start to write the `Shredos - How To Guide` in ascidoc markup.

AsciiDoc markup is used by buildroot which is the software ShredOS is build upon. As an example, the buildroot asciidoc manual can be found [here](https://github.com/buildroot/buildroot/tree/master/docs/manual) . The resultant PDF document derived from the AsciiDoc markup can be found [here](https://buildroot.org/downloads/manual/manual.pdf) 

By doing it like this, hopefully users will be able to submit pull requests and add more information, like a user guide, make corrections, etc. Also asciidoc can output the raw text in multiple formats, html & pdf in particular.

I'll start the manual off with a `how to build shredos from source section`. Further sections would explain nwipe in more detail.

Don't expect too much in regards to my spelling and grammar. I'm not a professional technical document writer (just an engineer). I'll be writing this pretty fast and I'm hoping others will be able to correct any grammatical errors they come across.

>And i'm a newbie in compiling... 

Have you tried building nwipe from source first? The procedure for that is pretty straight forward and I even wrote a script that will install all the dependencies, download the latest nwipe, build it and run it with a single command. It will give you a grounding in pretty much how every bit of C/C++ software is compiled.
[Nwipe - Compiling and installation](https://github.com/martijnvanbrummelen/nwipe#compiling--installing)
[Nwipe - Automating the download and compilation process](https://github.com/martijnvanbrummelen/nwipe#automating-the-download-and-compilation-process-for-debian-based-distros)

If you haven't already compiled nwipe from source, then I'd try that first. If you get stuck just let me know. Building shredos is far more complicated than building nwipe, not that I'm trying to put you off, everything is easy once you know how.. In fact taking a vanilla buildroot and adding the shredos stuff, then going through the literally thousands of settings to make sure everything is as it should be can take days if not weeks. However, it's nowhere nearly as bad as that if you `git clone` the shredos source where I've done all the tedious work of going through all the thousands of possibilities in `make linux-menuconfig`,`make busybox menuconfig` and `make menuconfig`. Currently, I'm building shredos on the 5.13 kernel and I'm now on day 3. Much of the time is spent double checking no drivers or applications have been left out.

Also I use ubuntu (KDE Neon) so if your using anything else that's not Debian based commands will probably differ to some degree.

**Reply by tlschenkjr** [reply]
Did this document every see the light of day?  This thread is almost a year old and I still cannot locate any instructions on building an ISO image instead of the bootable USB image.  I have successfully built a custom bootable USB image, but am stuck on building a matching ISO.  My custom image modifies the grub configuration and also adds sftp instead of lftp as an option for log files.

**Reply by PartialVolume** [reply]
No, I was hoping somebody would take up the project of writing this user manual but no takers, I could write it myself but to be honest I prefer writing code than writing user manuals. So for me writing the user manual comes after adding the features that need adding.

In regards to building an iso from your existing .img build here's some notes I made. This are personal notes so it may assume things, that make sense to me but might not be clear for somebody else.

So you don't mess up your existing .img build, I copy the entire directory .img build directory to somewhere else and that becomes my .iso build. That is basically what steps 1 and 2 are all about.

I was reading that in the next buildroot release it becomes a bit easier to combine multiple builds so this procedure might change in future releases.

Building a legacy boot (not UEFI)  (Not completed, currently the .iso doesn't support UEFI)

Using make menuconfig..
1. IMPORTANT ! Check the directory in build options, change to the appropriate location of the buildroot you are actually working on. This is relevant for me when I'm copying the .img build to the .iso build in a different directory. 
2. Select Build Options, then change the directory from /media/jbloggs/Data/Software/shredos/Shredos.2020.05/shredos.2020.02/configs/shredos_defconf  to /media/jbloggs/Data/Software/shredos/Shredos.2020.05.iso/shredos.2020.02/configs/shredos_defconf , i.e. the only difference is appending .iso so the path becomes  ../Shredos.2020.05.iso/..
3. Select bootloaders->Platform-> i386-pc (old bios boot method)
4. Change Boot Partition from 'hd0,msdos1' to 'cd'
5. Select bootloaders->builtin modules include biosdisk & iso9660
6. Select Filesystem Image->iso image-> grub2
7. Save and then exit make menuconfig.
8. cp board/shredos/grub.cfg  fs/iso9660/grub.cfg
9. Edit fs/iso9660/grub.cfg and replace /boot/shredos with /boot/bzImage (IMPORTANT) Get the case correct, that's a capital 'I'  not 'i' in bzImage.
10. Make sure the 32 bit futex fix is added on 32 bit architecture. CONFIG_COMPAT_32BIT_TIME see keep note.
11. make savedefconfig
12. make clean
13. make shredos_defconfig
14. make
15. Once completed and with no errors. cd output/images
16.  cp iso9660 to the name of the .img file but instead of .img use the .iso extension.

Building a UEFI (ignore these next lines, I never got around to adding UEFI to .iso and you don't need these for a legacy bios .iso build)
1. Select bootloaders->Platform-> i386-efi
2. Boot Partition option will disappear, so don't worry about that.
3. Select bootloaders->builtin modules REMOVE biosdisk & INCLUDE iso9660
4. cp ../shredos/grub.cfg  ../fs.iso9660/grub.cfg
5.  to be completed .....

**Reply by PartialVolume** [reply]
If you are reading the previous comment in an email, please read in a browser as I've made an number of edits that won't be reflected in the email github sends out. https://github.com/PartialVolume/shredos.x86_64/discussions/83#discussioncomment-4257263

**Reply by tlschenkjr** [reply]
Thanks for the response.  I'm trying out your instructions now.

**Reply by PartialVolume** [reply]
No problem, in regards to step 10, it mentions a keep note. This is the note. This should already be enabled in the config as you are copying your original .img build, but it's worth checking. Here's how:

https://forums.gentoo.org/viewtopic-t-1111448-start-0.html

According to the URL linked above.

This error "futex returns unexpected error code" disappears after I select CONFIG_COMPAT_32BIT_TIME

CONFIG_COMPAT_32BIT_TIME:

This enables 32 bit time_t support in addition to 64 bit time_t support.
This is relevant on all 32-bit architectures, and 64-bit architectures as part of compat syscall handling.

Symbol: COMPAT_32BIT_TIME [=y]
Type   : bool
Defined at arch/Kconfig:837
  Prompt: Provide system calls for 32-bit time_t
  Location:
    -> General architecture-dependent options

**Reply by PartialVolume** [reply]
There is a issue #78 that explains what the original issue "futex returns unexpected error code" was all about and the fix.

**Comment by Firminator** [comment]
:) This should go into https://github.com/PartialVolume/shredos.x86_64/discussions/categories/q-a, too, since it's not an issue, but a question. One of these days I'll have to re-try compiling ShredOS again and see how far I come.
